astropy/io/fits/hdu/base.py

```diff
@@ -1356,11 +1356,11 @@ def _calculate_checksum(self, datasum, checksum_keyword='CHECKSUM'):
         old_checksum = self._header[checksum_keyword]
         self._header[checksum_keyword] = '0' * 16
 
-        # Convert the header to a string.
-        s = str(self._header)
+        # Convert the header to bytes.
+        s = self._header.tostring().encode('utf8')
```

Comment:
I thinks it's impossible to pass in non-ascii (or non-latin-1) characters in a Header. However the binary representation of these should be the same as for utf-8 so there's (probably) no problem.